File: src/bridge.ts
/**
 * OpenClaw Discord Bridge - Standalone Application
 * 
 * A standalone bridge that connects Discord to OpenClaw via HTTP API.
 * This approach provides full proxy support for both WebSocket and REST API calls,
 * fixing issue #27409.
 * 
 * Architecture:
 *   Discord Bot (with proxy) → Message Router → OpenClaw HTTP API
 */

import { Client, GatewayIntentBits, Partials, Events, Message, TextChannel, DMChannel } from 'discord.js';
import { HttpsProxyAgent } from 'https-proxy-agent';
import { calculateIntents, resolveAccountConfig } from './client';
import type { AccountConfig, IntentsConfig, DiscordMessageContext } from './types';

/**
 * Bridge configuration from environment variables
 */
interface BridgeConfig {
  discordToken: string;
  proxyUrl?: string;
  openclawHttpUrl: string;
  openclawAuthToken: string;
  dmPolicy: 'open' | 'pairing' | 'allowlist' | 'disabled';
  groupPolicy: 'open' | 'allowlist' | 'disabled';
  allowFrom?: string[];
  intents?: IntentsConfig;
  deviceId: string;
}

/**
 * Session key for OpenClaw routing
 */
interface SessionKey {
  agentId: string;
  channel: 'discord';
  accountId: string;
  peer?: string;
  guildId?: string;
  channelId?: string;
}

/**
 * Message payload sent to OpenClaw HTTP API
 */
interface OpenClawMessagePayload {
  sessionKey: SessionKey;
  content: string;
  timestamp: number;
}

/**
 * Parse bridge configuration from environment variables
 */
function loadBridgeConfig(): BridgeConfig {
  const discordToken = process.env.DISCORD_TOKEN;
  const openclawAuthToken = process.env.OPENCLAW_AUTH_TOKEN;

  if (!discordToken) {
    throw new Error('DISCORD_TOKEN environment variable is required');
  }

  if (!openclawAuthToken) {
    throw new Error('OPENCLAW_AUTH_TOKEN environment variable is required');
  }

  return {
    discordToken,
    proxyUrl: process.env.PROXY_URL || 'http://127.0.0.1:7890',
    openclawHttpUrl: process.env.OPENCLAW_HTTP_URL || 'http://127.0.0.1:18789',
    openclawAuthToken,
    dmPolicy: (process.env.DM_POLICY as any) || 'open',
    groupPolicy: (process.env.GROUP_POLICY as any) || 'allowlist',
    allowFrom: process.env.ALLOW_FROM?.split(',').map(s => s.trim()),
    deviceId: process.env.DEVICE_ID || 'openclaw-discord-bridge',
    intents: {
      messageContent: process.env.INTENT_MESSAGE_CONTENT !== 'false',
      serverMembers: process.env.INTENT_SERVER_MEMBERS !== 'false',
      presence: process.env.INTENT_PRESENCE === 'true',
    },
  };
}

/**
 * Create a proxy-aware Discord client for the bridge
 * 
 * The key fix for issue #27409 is injecting the proxy agent into
 * both WebSocket and REST API calls.
 */
function createBridgeClient(config: BridgeConfig): Client {
  const intents = calculateIntents(config.intents);

  const clientOptions: ConstructorParameters<typeof Client>[0] = {
    intents,
    partials: [
      Partials.Channel,
      Partials.Message,
      Partials.Reaction,
    ],
  };

  // Inject proxy agent for both WebSocket and REST API
  if (config.proxyUrl) {
    const proxyAgent = new HttpsProxyAgent(config.proxyUrl);
    clientOptions.rest = {
      agent: proxyAgent as any,
    };
  }

  return new Client(clientOptions);
}

/**
 * Generate session key for OpenClaw routing
 */
function generateSessionKey(
  message: Message,
  accountId: string,
  deviceId: string
): SessionKey {
  const isDm = !message.guild;

  if (isDm) {
    return {
      agentId: 'default',
      channel: 'discord',
      accountId,
      peer: message.author.id,
    };
  }

  return {
    agentId: 'default',
    channel: 'discord',
    accountId,
    guildId: message.guild!.id,
    channelId: message.channel.id,
  };
}

/**
 * Send message to OpenClaw HTTP API
 */
async function sendToOpenClaw(
  config: BridgeConfig,
  payload: OpenClawMessagePayload
): Promise<void> {
  const url = `${config.openclawHttpUrl}/api/v1/messages/ingest`;
  
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${config.openclawAuthToken}`,
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`OpenClaw API error: ${response.status} ${errorText}`);
    }

    console.log(`[Bridge] Message sent to OpenClaw: ${payload.sessionKey.peer || payload.sessionKey.channelId}`);
  } catch (error) {
    console.error('[Bridge] Failed to send message to OpenClaw:', error);
    throw error;
  }
}

/**
 * Check if user is allowed to send messages
 */
function isUserAllowed(
  userId: string,
  config: BridgeConfig,
  guildId?: string
): boolean {
  // Check DM policy
  if (!guildId) {
    switch (config.dmPolicy) {
      case 'open':
        return true;
      case 'disabled':
        return false;
      case 'allowlist':
        return config.allowFrom?.includes(userId) ?? false;
      case 'pairing':
      default:
        // For simplicity, pairing mode allows all in standalone bridge
        return true;
    }
  }

  // Guild message policy
  if (config.groupPolicy === 'disabled') {
    return false;
  }

  return true;
}

/**
 * Check if bot should respond to a message
 */
function shouldRespond(
  message: Message,
  config: BridgeConfig
): boolean {
  // Ignore bot messages
  if (message.author.bot) {
    return false;
  }

  const guildId = message.guild?.id;
  const userId = message.author.id;

  // Check if user is allowed
  if (!isUserAllowed(userId, config, guildId)) {
    return false;
  }

  // For guild messages, require mention
  if (guildId) {
    const mentioned = message.mentions.users?.has(message.client.user?.id ?? '');
    if (!mentioned) {
      return false;
    }
  }

  return true;
}

/**
 * Send message to Discord channel with proxy support
 */
async function sendDiscordMessage(
  client: Client,
  channelId: string,
  content: string
): Promise<void> {
  const channel = await client.channels.fetch(channelId);
  
  if (!channel) {
    throw new Error(`Channel ${channelId} not found`);
  }

  if (!('send' in channel)) {
    throw new Error(`Channel ${channelId} is not text-based`);
  }

  await (channel as TextChannel | DMChannel).send({ content });
}

/**
 * Main bridge class
 */
class DiscordBridge {
  private client: Client;
  private config: BridgeConfig;
  private accountId: string = 'default';

  constructor(config: BridgeConfig) {
    this.config = config;
    this.client = createBridgeClient(config);
    this.setupEventHandlers();
  }

  private setupEventHandlers(): void {
    // Handle client ready event
    this.client.once(Events.ClientReady, async () => {
      console.log(`[Bridge] Connected as ${this.client.user?.tag}`);
      console.log(`[Bridge] Proxy: ${this.config.proxyUrl || 'none'}`);
      console.log(`[Bridge] OpenClaw API: ${this.config.openclawHttpUrl}`);
      
      // Set presence if desired
      await this.client.user?.setPresence({
        status: 'online',
        activities: [{ name: 'OpenClaw Bridge', type: 2 }], // Type 2 = Listening
      });
    });

    // Handle incoming messages
    this.client.on(Events.MessageCreate, async (message) => {
      if (!shouldRespond(message, this.config)) {
        return;
      }

      try {
        await this.handleMessage(message);
      } catch (error) {
        console.error('[Bridge] Error handling message:', error);
      }
    });

    // Handle errors
    this.client.on(Events.Error, (error) => {
      console.error('[Bridge] Discord error:', error);
    });
  }

  private async handleMessage(message: Message): Promise<void> {
    console.log(`[Bridge] Message from ${message.author.tag} in ${message.channel.id}`);

    const sessionKey = generateSessionKey(
      message,
      this.accountId,
      this.config.deviceId
    );

    const payload: OpenClawMessagePayload = {
      sessionKey,
      content: message.content,
      timestamp: Date.now(),
    };

    await sendToOpenClaw(this.config, payload);
  }

  /**
   * Start the bridge
   */
  async start(): Promise<void> {
    console.log('[Bridge] Starting Discord Bridge...');
    await this.client.login(this.config.discordToken);
  }

  /**
   * Stop the bridge
   */
  async stop(): Promise<void> {
    console.log('[Bridge] Stopping Discord Bridge...');
    await this.client.destroy();
  }
}

/**
 * Main entry point
 */
async function main(): Promise<void> {
  try {
    const config = loadBridgeConfig();
    
    console.log('[Bridge] Configuration loaded:');
    console.log(`  - Discord Token: ${config.discordToken.substring(0, 10)}...`);
    console.log(`  - Proxy URL: ${config.proxyUrl || 'none'}`);
    console.log(`  - OpenClaw URL: ${config.openclawHttpUrl}`);
    console.log(`  - DM Policy: ${config.dmPolicy}`);
    console.log(`  - Group Policy: ${config.groupPolicy}`);
    console.log(`  - Device ID: ${config.deviceId}`);

    const bridge = new DiscordBridge(config);
    await bridge.start();

    // Handle graceful shutdown
    process.on('SIGINT', async () => {
      console.log('\n[Bridge] Received SIGINT, shutting down...');
      await bridge.stop();
      process.exit(0);
    });

    process.on('SIGTERM', async () => {
      console.log('\n[Bridge] Received SIGTERM, shutting down...');
      await bridge.stop();
      process.exit(0);
    });

  } catch (error) {
    console.error('[Bridge] Failed to start:', error);
    process.exit(1);
  }
}

// Run the bridge
main();

File: bridge.py
"""
OpenClaw Discord Bridge - Standalone Application

A standalone bridge that connects Discord to OpenClaw via HTTP API.
This approach provides full proxy support for both WebSocket and REST API calls,
fixing issue #27409.

Architecture:
    Discord Bot (with proxy) → Message Router → OpenClaw HTTP API

Usage:
    python bridge.py
    
Or with environment variables:
    DISCORD_TOKEN="your_token" OPENCLAW_AUTH_TOKEN="your_token" python bridge.py
"""

import asyncio
import logging
from datetime import datetime
from typing import Optional, Dict, Any

import discord
from discord import Intents, Message, TextChannel, DMChannel, PartialMessageable
from aiohttp import ClientSession, ClientTimeout, TCPConnector
from aiohttp_socks import ProxyConnector

from config import load_bridge_config, print_config_summary
from types import (
    BridgeConfig,
    SessionKey,
    OpenClawMessagePayload,
    DMPolicy,
    GroupPolicy,
)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(__name__)


def generate_session_key(
    message: Message,
    account_id: str,
    device_id: str
) -> SessionKey:
    """
    Generate session key for OpenClaw routing.
    
    Args:
        message: The Discord message
        account_id: The account identifier
        device_id: The device identifier
        
    Returns:
        SessionKey for routing
    """
    is_dm = not message.guild
    
    if is_dm:
        return SessionKey(
            agent_id="default",
            channel="discord",
            account_id=account_id,
            peer=message.author.id,
        )
    
    return SessionKey(
        agent_id="default",
        channel="discord",
        account_id=account_id,
        guild_id=message.guild.id,
        channel_id=message.channel.id,
    )


def create_proxy_connector(proxy_url: Optional[str]) -> Optional[ProxyConnector]:
    """
    Create a proxy connector for aiohttp.
    
    Args:
        proxy_url: The proxy URL (e.g., "http://127.0.0.1:7890")
        
    Returns:
        ProxyConnector or None if no proxy
    """
    if not proxy_url:
        return None
    
    # Parse proxy URL and create appropriate connector
    # aiohttp_socks handles http, https, socks4, socks5
    try:
        return ProxyConnector.from_url(proxy_url)
    except ValueError as e:
        logger.warning(f"Failed to create proxy connector: {e}")
        return None


async def send_to_openclaw(
    session: ClientSession,
    config: BridgeConfig,
    payload: OpenClawMessagePayload
) -> None:
    """
    Send message to OpenClaw HTTP API.
    
    Args:
        session: The aiohttp client session
        config: Bridge configuration
        payload: Message payload to send
        
    Raises:
        Exception: If the request fails
    """
    url = f"{config.openclaw_http_url}/api/v1/messages/ingest"
    
    data = {
        "sessionKey": {
            "agentId": payload.session_key.agent_id,
            "channel": payload.session_key.channel,
            "accountId": payload.session_key.account_id,
        },
        "content": payload.content,
        "timestamp": payload.timestamp,
    }
    
    # Add optional fields
    if payload.session_key.peer:
        data["sessionKey"]["peer"] = payload.session_key.peer
    if payload.session_key.guild_id:
        data["sessionKey"]["guildId"] = payload.session_key.guild_id
    if payload.session_key.channel_id:
        data["sessionKey"]["channelId"] = payload.session_key.channel_id
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {config.openclaw_auth_token}",
    }
    
    try:
        async with session.post(url, json=data, headers=headers) as response:
            if response.status != 200:
                error_text = await response.text()
                raise Exception(f"OpenClaw API error: {response.status} {error_text}")
            
            peer_id = payload.session_key.peer or payload.session_key.channel_id
            logger.info(f"Message sent to OpenClaw: {peer_id}")
            
    except Exception as e:
        logger.error(f"Failed to send message to OpenClaw: {e}")
        raise


def is_user_allowed(
    user_id: str,
    config: BridgeConfig,
    guild_id: Optional[str] = None
) -> bool:
    """
    Check if user is allowed to send messages.
    
    Args:
        user_id: The Discord user ID
        config: Bridge configuration
        guild_id: The guild ID (None for DMs)
        
    Returns:
        True if user is allowed, False otherwise
    """
    # Check DM policy
    if not guild_id:
        match config.dm_policy:
            case DMPolicy.OPEN:
                return True
            case DMPolicy.DISABLED:
                return False
            case DMPolicy.ALLOWLIST:
                return user_id in config.allow_from
            case DMPolicy.PAIRING:
                # For simplicity, pairing mode allows all in standalone bridge
                return True
    
    # Guild message policy
    if config.group_policy == GroupPolicy.DISABLED:
        return False
    
    return True


def should_respond(
    message: Message,
    config: BridgeConfig
) -> bool:
    """
    Check if bot should respond to a message.
    
    Args:
        message: The Discord message
        config: Bridge configuration
        
    Returns:
        True if bot should respond, False otherwise
    """
    # Ignore bot messages
    if message.author.bot:
        return False
    
    guild_id = message.guild.id if message.guild else None
    user_id = message.author.id
    
    # Check if user is allowed
    if not is_user_allowed(user_id, config, guild_id):
        return False
    
    # For guild messages, require mention
    if guild_id:
        if not message.mentions or not any(
            member.id == message.guild.me.id for member in message.mentions
        ):
            return False
    
    return True


class DiscordBridge:
    """
    Main bridge class connecting Discord to OpenClaw.
    """
    
    def __init__(self, config: BridgeConfig):
        """
        Initialize the Discord bridge.
        
        Args:
            config: Bridge configuration
        """
        self.config = config
        self.session: Optional[ClientSession] = None
        self.account_id = "default"
        
        # Create Discord intents
        intents = Intents.default()
        intents.message_content = config.intents.message_content
        intents.members = config.intents.server_members
        intents.presence = config.intents.presence
        
        # Create Discord client with proxy support
        proxy_url = config.proxy_url
        self.client = discord.Client(
            intents=intents,
            proxy=proxy_url,
        )
        
        # Setup event handlers
        self.client.event(self.on_ready)
        self.client.event(self.on_message)
        self.client.event(self.on_error)
    
    async def on_ready(self) -> None:
        """Handle client ready event."""
        logger.info(f"Connected as {self.client.user}")
        logger.info(f"Proxy: {self.config.proxy_url or 'none'}")
        logger.info(f"OpenClaw API: {self.config.openclaw_http_url}")
        
        # Set presence
        if self.config.activity:
            activity = discord.Activity(
                name=self.config.activity,
                type=self.config.activity_type.value,
                url=self.config.activity_url,
            )
            await self.client.change_presence(
                status=discord.Status[self.config.status.value.upper()],
                activity=activity,
            )
    
    async def on_message(self, message: Message) -> None:
        """
        Handle incoming messages.
        
        Args:
            message: The received message
        """
        if not should_respond(message, self.config):
            return
        
        try:
            await self.handle_message(message)
        except Exception as e:
            logger.error(f"Error handling message: {e}", exc_info=True)
    
    async def on_error(self, event: str, *args: Any, **kwargs: Any) -> None:
        """
        Handle Discord client errors.
        
        Args:
            event: The event name
            *args: Event arguments
            **kwargs: Event keyword arguments
        """
        logger.error(f"Discord error in {event}: {args}", exc_info=True)
    
    async def handle_message(self, message: Message) -> None:
        """
        Process and forward a message to OpenClaw.
        
        Args:
            message: The Discord message to process
        """
        logger.info(f"Message from {message.author} in channel {message.channel.id}")
        
        session_key = generate_session_key(
            message,
            self.account_id,
            self.config.device_id
        )
        
        payload = OpenClawMessagePayload(
            session_key=session_key,
            content=message.content,
            timestamp=int(datetime.now().timestamp() * 1000),
        )
        
        if self.session:
            await send_to_openclaw(self.session, self.config, payload)
    
    async def start(self) -> None:
        """Start the Discord bridge."""
        logger.info("Starting Discord Bridge...")
        
        # Create aiohttp session with proxy
        connector = create_proxy_connector(self.config.proxy_url)
        timeout = ClientTimeout(total=30)
        
        self.session = ClientSession(
            connector=connector,
            timeout=timeout,
        )
        
        # Start Discord client
        await self.client.start(self.config.discord_token)
    
    async def stop(self) -> None:
        """Stop the Discord bridge."""
        logger.info("Stopping Discord Bridge...")
        
        if self.session:
            await self.session.close()
        
        if self.client.is_ready():
            await self.client.close()


async def main() -> None:
    """Main entry point."""
    try:
        config = load_bridge_config()
        print_config_summary(config)
        
        bridge = DiscordBridge(config)
        
        # Handle graceful shutdown
        import signal
        
        async def shutdown(signum):
            logger.info(f"Received signal {signum}, shutting down...")
            await bridge.stop()
        
        loop = asyncio.get_running_loop()
        for sig in (signal.SIGINT, signal.SIGTERM):
            loop.add_signal_handler(sig, lambda s=sig: asyncio.create_task(shutdown(s)))
        
        await bridge.start()
        
    except ValueError as e:
        logger.error(f"Configuration error: {e}")
        exit(1)
    except discord.LoginFailure:
        logger.error("Invalid Discord token")
        exit(1)
    except Exception as e:
        logger.error(f"Failed to start: {e}", exc_info=True)
        exit(1)


if __name__ == "__main__":
    asyncio.run(main())